src/crypto/tls/delegated_credentials.go

```diff
@@ -0,0 +1,533 @@
+	}
+
+	sigAlgo, err := getSigAlgo(cert)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// Generate the Delegated Credential Key Pair based on the provided scheme
+	var privK crypto.PrivateKey
+	var pubK crypto.PublicKey
+	switch pubAlgo {
+	case ECDSAWithP256AndSHA256,
+		ECDSAWithP384AndSHA384,
+		ECDSAWithP521AndSHA512:
+		privK, err = ecdsa.GenerateKey(getCurve(pubAlgo), rand.Reader)
+		if err != nil {
+			return nil, nil, err
+		}
+		pubK = privK.(*ecdsa.PrivateKey).Public()
+	case Ed25519:
+		pubK, privK, err = ed25519.GenerateKey(rand.Reader)
+		if err != nil {
+			return nil, nil, err
+		}
+	default:
+		return nil, nil, fmt.Errorf("tls: unsupported algorithm for delegated credential: %T", pubAlgo)
+	}
+
+	// Prepare the credential for signing
+	hash := getHash(sigAlgo)
+	credential := &Credential{validTime, pubAlgo, pubK}
+	values, err := prepareDelegationSignatureInput(hash, credential, cert.Leaf.Raw, sigAlgo, peer)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	var sig []byte
+	switch sk := cert.PrivateKey.(type) {
+	case *ecdsa.PrivateKey:
+		opts := crypto.SignerOpts(hash)
+		sig, err = sk.Sign(rand.Reader, values, opts)
+		if err != nil {
+			return nil, nil, err
+		}
+	case ed25519.PrivateKey:
+		opts := crypto.SignerOpts(hash)
+		sig, err = sk.Sign(rand.Reader, values, opts)
+		if err != nil {
+			return nil, nil, err
+		}
+	default:
+		return nil, nil, fmt.Errorf("tls: unsupported key type for delegated credential")
+	}
+
+	return &DelegatedCredential{
+		Cred:      credential,
+		Algorithm: sigAlgo,
+		Signature: sig,
+	}, privK, nil
+}
+
+// Validate checks that the delegated credential is valid by checking that the
+// signature is valid, that the credential hasn't expired, and that the TTL is
+// valid. It also checks that certificate can be used for delegation.
+func (dc *DelegatedCredential) Validate(cert *x509.Certificate, peer string, now time.Time, certVerifyMsg *certificateVerifyMsg) bool {
+	if dc.IsExpired(cert.NotBefore, now) {
+		return false
+	}
+
+	if dc.InvalidTTL(cert.NotBefore, now) {
+		return false
+	}
+
+	if !isValidInCertVerify(dc.Cred.expCertVerfAlgo, certVerifyMsg.signatureAlgorithm) {
+		return false
+	}
+
+	if !isValidForDelegation(cert) {
+		return false
+	}
+
+	hash := getHash(dc.Algorithm)
+	in, err := prepareDelegationSignatureInput(hash, dc.Cred, cert.Raw, dc.Algorithm, peer)
+	if err != nil {
+		return false
+	}
+
+	switch dc.Algorithm {
+	case ECDSAWithP256AndSHA256,
+		ECDSAWithP384AndSHA384,
+		ECDSAWithP521AndSHA512:
+		pk, ok := cert.PublicKey.(*ecdsa.PublicKey)
+		if !ok {
+			return false
+		}
+
+		return ecdsa.VerifyASN1(pk, in, dc.Signature)
+	case Ed25519:
+		pk, ok := cert.PublicKey.(ed25519.PublicKey)
+		if !ok {
+			return false
+		}
+
+		return ed25519.Verify(pk, in, dc.Signature)
+	default:
+		return false
+	}
+}
+
+// Marshal encodes a DelegatedCredential structure. It also sets dc.Raw to the encoding.
+func (dc *DelegatedCredential) Marshal() ([]byte, error) {
```

Comment:
Using MarshalBinary would be more idiomatic.

Reply:
(see encoding/BinaryMarshaler)

Reply:
In a way. But I think this relates more to the way it is defined on `handshake_messages.go` which use Marshal.

Reply:
But that's all private.  DelegatedCredential is a public structure and your `Marshal` function is all but in name following `encoding.BinaryMarshaler`.

Reply:
This one is private now as well.